replication-api/src/main/java/org/codice/ditto/replication/api/modern/ReplicationPersistentStore.java

```diff
@@ -0,0 +1,55 @@
+/**
+ * Copyright (c) Connexta
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ditto.replication.api.modern;
+
+import java.util.List;
+import java.util.Optional;
+
+/** A ReplicationPersistentStore performs CRUD operations for replication. */
+public interface ReplicationPersistentStore {
+
+  /**
+   * Attempts to find an object that has been saved with the given persistence type and ID. It then
+   * creates an instance of class {@code <T>} with the data retrieved from persistence.
+   *
+   * @param klass The class of object to instantiate and return.
+   * @param typeName The persistence type of the desired object.
+   * @param id The id of the object to search for.
+   * @param <T> A class that implements the Persistable interface. An object of this type will be
+   *     created from the persisted data and returned.
+   * @return The object that was retrieved from storage
+   */
+  <T extends Persistable> Optional<T> get(Class<T> klass, String typeName, String id);
+
+  /**
+   * Attempts to retrieve all objects saved with the given persistence type. It then creates a list
+   * of objects of class {@code <T>} with the data retrieved from persistence.
+   *
+   * @param klass The class of object to instantiate and return.
+   * @param typeName The persistence type of the desired object.
+   * @param <T> A class that implements the Persistable interface. Objects of this type will be
+   *     created from the persisted data and returned.
+   * @return A list of Objects that were retrieved from storage
+   */
+  <T extends Persistable> List<T> get(Class<T> klass, String typeName);
```

Comment:
❓ How about returning a stream instead and let the client decide if they actually need a collection or not? This avoids creating temporary collections for no reasons.

Reply:
☝️ Same comment as before with respect to type and class